
This patch modifies how Qt reports DRM devices on non-PCI platforms.

On non-PCI platforms vanilla Qt reports DRM devices with order of uDEV
reports them. On platforms with discrete GPU & DE (display engine)
kernel modules init concurrency results that minor dev numbers are not
peristant across reboots (on one boot GPU & DE are /dev/dri/card0 and /dev/dri/card1
on anoter boot they are reversed). This makes issue with apps crashing at
start when Qt elects GPU DRM devices as screen DRM device. Using Qt KMS json
config to declare what DRM device is screen DRM device not helps as DRM devices
minors are not persisten across boots. Only solution seems to be to realy on uDEV
attribute allowing to identify DE DRM device: PrimaryGPU & assuming DE kernel module.
properly sets this attribute.

This patch makes Qt reporting as first DRM device on list of connected DRM devices, an
device with PrimaryGPU attribute. If kernel driver not sets PrimaryGPU attribute - old
(uDEV order) is maintained


diff -Naur qt-everywhere-src-5.15.2-old/qtbase/src/platformsupport/devicediscovery/qdevicediscovery_udev.cpp qt-everywhere-src-5.15.2-new/qtbase/src/platformsupport/devicediscovery/qdevicediscovery_udev.cpp
--- qt-everywhere-src-5.15.2-old/qtbase/src/platformsupport/devicediscovery/qdevicediscovery_udev.cpp	2020-10-27 09:02:11.000000000 +0100
+++ qt-everywhere-src-5.15.2-new/qtbase/src/platformsupport/devicediscovery/qdevicediscovery_udev.cpp	2021-11-30 10:42:00.226551847 +0100
@@ -107,6 +107,7 @@
 QStringList QDeviceDiscoveryUDev::scanConnectedDevices()
 {
     QStringList devices;
+    QStringList devices_sec_priority;
 
     if (!m_udev)
         return devices;
@@ -146,17 +147,23 @@
             if (m_types & Device_DRM_PrimaryGPU) {
                 udev_device *pci = udev_device_get_parent_with_subsystem_devtype(udevice, "pci", 0);
                 if (pci) {
-                    if (qstrcmp(udev_device_get_sysattr_value(pci, "boot_vga"), "1") == 0)
+                    if (qstrcmp(udev_device_get_sysattr_value(pci, "boot_vga"), "1") == 0) {
+                        qCDebug(lcDD) << "Found PrimaryGPU DRM device:" << devices;
                         devices << candidate;
+                    }
+                } else {
+                    qCDebug(lcDD) << "Found PrimaryGPU DRM device:" << devices;
+                    devices << candidate;
                 }
             } else
-                devices << candidate;
+                devices_sec_priority << candidate;
         }
 
         udev_device_unref(udevice);
     }
     udev_enumerate_unref(ue);
 
+    devices << devices_sec_priority;
     qCDebug(lcDD) << "Found matching devices" << devices;
 
     return devices;
